### PR TITLE
Feature/add published check to get population types

### PR DESCRIPTION
**Pull request by ChrysmOre**
### What

Added check to /population-types to ensure we only list population types that are published (i.e. have at least one published dataset associated with them)

This is achieved by checking at least one dataset that is based off the population type has a 'current' document.

This PR includes refactoring of existing component tests and removal of redundant unit tests to keep inline with our common practices.

### How to review

Check changes make sense, tests pass

### Who can review

Anyone

### CODE REVIEW

File: service/interfaces.go [hunk 1]
```diff
@@ -5,21 +5,24 @@ import (
...
+//go:generate moq -out mock/cantabular_client.go -pkg mock . CantabularClient
+//go:generate moq -out mock/dataset_api_client.go -pkg mock . DatasetAPIClient
 
 // Initialiser defines the methods to initialise external services
 type Initialiser interface {
 	GetHTTPServer(bindAddr string, router http.Handler) HTTPServer
 	GetCantabularClient(cfg config.CantabularConfig) CantabularClient
 	GetHealthCheck(cfg *config.Config, time, commit, version string) (HealthChecker, error)
 	GetResponder() Responder
+	GetDatasetAPIClient(cfg *config.Config) DatasetAPIClient
```
Comment (saminahbab) [inline service/interfaces.go:25]: I have put in a lot of this infrastructure work also, have we repeated work? looks like a lot of work has been duplicated?

File: handler/population_types.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+				break
+			}
+		}
+		if isPublished {
+			published = append(published, p)
+		}
+	}
+
+	if len(published) == 0 {
+		h.respond.Error(ctx, w, http.StatusNotFound, errors.New("no population types found"))
```
Comment (saminahbab) [inline handler/population_types.go:82]: would prefer breaking splitting this by line...cleaner

File: handler/population_types.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	req := cantabular.GetGeographyDimensionsRequest{
+		Dataset: chi.URLParam(r, "population-type"),
+	}
+
+	res, err := h.cantabular.GetGeographyDimensions(ctx, req)
+	if err != nil {
+		h.respond.Error(
+			ctx,
+			w,
+			h.cantabular.StatusCode(err), // Can be changed to ctblr.StatusCode(err) once added to Client
```
Comment (saminahbab) [inline handler/population_types.go:106]: remove comment

File: service/initialise.go [hunk 4]
```diff
@@ -3,13 +3,15 @@ package service
...
+	"github.com/ONSdigital/dp-api-clients-go/v2/dataset"
 	"github.com/ONSdigital/dp-healthcheck/healthcheck"
 	dphttp "github.com/ONSdigital/dp-net/http"
 	"github.com/ONSdigital/dp-net/v2/responder"
-	"github.com/ONSdigital/dp-population-types-api/config"
+
+	"github.com/pkg/errors"
 )
 
 // Init implements the Initialiser interface to initialise dependencies
```
Comment (saminahbab) [inline service/initialise.go:17]: redundant comment, not yours but best to remove bloat

File: service/routes.go [hunk 5]
```diff
@@ -22,7 +22,7 @@ func (svc *Service) publicEndpoints(ctx context.Context) {
 	log.Info(ctx, "enabling public endpoints")
 
 	// Routes
-	populationTypes := handler.NewPopulationTypes(svc.responder, svc.cantabularClient)
+	populationTypes := handler.NewPopulationTypes(svc.Config, svc.responder, svc.cantabularClient, svc.datasetAPIClient)
```
Comment (saminahbab) [inline service/routes.go:25]: would rather break each var than only half the vars

File: service/service.go [hunk 6]
```diff
@@ -105,8 +107,7 @@ func (svc *Service) Close(ctx context.Context) error {
 	return nil
 }
 
-func (svc *Service) registerCheckers(ctx context.Context) (err error) {
-
+func (svc *Service) registerCheckers(ctx context.Context) error {
 	if svc.cantabularClient != nil {
 		if svc.Config.CantabularHealthcheckEnabled {
```
Comment (saminahbab) [inline service/service.go:112]: dataset checker should probably be added here? I added it so probs just can wait for that

File: service/routes_test.go [hunk 7]
```diff
@@ -1,69 +0,0 @@
-package service_test
```
Comment (saminahbab) [inline service/routes_test.go:1]: personally quite down to remove the unit tests, I feel that a lot of them are bloat... but big deletions like these should get eyes from the wider team also

File: service/interfaces.go [hunk 8]
```diff
@@ -51,3 +54,8 @@ type Responder interface {
 	StatusCode(http.ResponseWriter, int)
 	Bytes(context.Context, http.ResponseWriter, int, []byte)
 }
+
+type DatasetAPIClient interface {
+	GetDatasets(ctx context.Context, uToken, svcToken, collectionID string, params *dataset.QueryParams) (dataset.List, error)
+	Checker(ctx context.Context, state *healthcheck.CheckState) error
```
Comment (saminahbab) [inline service/interfaces.go:60]: I see the definition of the checker as part of the contract, but it not being used I think

File: Makefile [hunk 9]
```diff
@@ -15,7 +15,6 @@ audit:
 
 .PHONY: lint
 lint:
-	golangci-lint run ./... --timeout 10m --skip-dirs=features --verbose --enable=gosec --enable=gocritic --enable=gofmt --enable=gocyclo --enable=bodyclose --enable=gocognit --enable=goimports
```
Comment (saminahbab) [inline Makefile:18]: why are we taking linting out of CI?

File: reflex [hunk 10]
```diff
@@ -1,3 +1,4 @@
 # Watch all files ending in .go, excluding files in `.go` directory or files
 # ending in `_test.go` & run `make debug` when any matching file is changed
 -s -r \.go$ -R ^\.go/ -R _test\.go$ make run
+
```
Comment (saminahbab) [inline reflex:4]: can remove empty line

File: service/interfaces.go (hunk 1 above)
Comment (ChrysmOre) [reply service/interfaces.go:25]: Not a problem, will resolve when rebasing

File: handler/population_types.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+				err,
+				"failed to get datasets",
+			))
+			return
+		}
+
+		log.Info(ctx, "datasets found", log.Data{"datasets": datasets})
+
+		var isPublished bool
+		for _, d := range datasets.Items {
```
[the remark this answers is not in the document]
Comment (ChrysmOre) [reply handler/population_types.go:70]: Yeah that's why I thought of the auth method.. I went ahead and used this original method when I wasn't getting the results I expected before I realised we had dud test data.. I will perhaps revert to the auth method

File: service/routes.go (hunk 5 above)
Comment (ChrysmOre) [reply service/routes.go:25]: Not sure what you mean?

File: Makefile (hunk 9 above)
Comment (ChrysmOre) [reply Makefile:18]: It was causing PRs to fail - Not from failing liniting but failing to download the golangci-lint itself.. And it was never discussed or agreed that it should be added in the first place. None of the other services have any linting running in CI.

File: handler/population_types.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ChrysmOre) [reply handler/population_types.go:70]: But also it's not going to be ridiculous like the 100,000s, will be in the hundreds at most which isn't much in those terms

[the remark this answers is not in the document]
Comment (saminahbab) [reply handler/population_types.go:70]: that still does take a non trivial amount of time.. happy to go with it for now, if others are, but probably best to go back to this and ensure that it is efficient enough

**Review by saminahbab — APPROVED** [review]

**Comment by saminahbab** [comment]
i approved too early, failing unit tests

**Review by saminahbab — APPROVED** [review]

**Pull request merged**